api/services/User/index.js

```diff
@@ -38,48 +38,88 @@ export default class User extends BaseModelService {
     }
   }
 
+  static doesKeyBelongToInvite(key, invite) {
+    if (key.user_id !== invite.user_id) {
+      return false;
+    }
+
+    const { excludedEvents = [] } = key;
+    if (excludedEvents.includes(invite.event.id)) {
+      return false;
+    }
+
+    if (
+      invite.event.start_at >= key.dateTimeRange.start_at &&
+      invite.event.start_at <= key.dateTimeRange.end_at
+    ) {
+      return true;
+    }
+
+    if (
+      invite.event.end_at >= key.dateTimeRange.start_at &&
+      invite.event.end_at <= key.dateTimeRange.end_at
+    ) {
+      return true;
+    }
+
+    if (
+      invite.event.start_at <= key.dateTimeRange.start_at &&
+      invite.event.end_at >= key.dateTimeRange.end_at
+    ) {
+      return true;
+    }
+
+    return false;
+  }
+
   static async findAllAvailabilityDuring(keys) {
-    return Promise.all(
-      keys.map(
-        async (key) =>
-          !(await Models.CalendarEventInvite.scope('visible').findOne({
+    return Models.CalendarEventInvite.scope('visible')
+      .findAll({
+        include: [
+          {
+            model: Models.CalendarEvent,
+            as: 'event',
             where: {
-              user_id: key.user_id,
-            },
-            include: [
-              {
-                model: Models.CalendarEvent,
-                as: 'event',
-                where: {
+              [Models.Sequelize.Op.or]: keys.map(
+                ({ user_id, dateTimeRange, excludeEvents }) => ({
                   [Models.Sequelize.Op.and]: [
-                    overlapDateTimeClause(key.dateTimeRange),
+                    {
+                      $user_id$: user_id,
+                    },
                     {
                       id: {
-                        [Models.Sequelize.Op.notIn]: key.exculdeEvents || [],
+                        [Models.Sequelize.Op.notIn]: excludeEvents || [],
                       },
                     },
+                    overlapDateTimeClause(dateTimeRange),
                   ],
-                },
-                required: true,
-              },
-            ],
-          })),
-      ),
-    );
+                }),
+              ),
+            },
+            required: true,
+          },
+        ],
+      })
+      .then((invites) =>
+        keys.map(
+          (key) =>
+            !invites.find((invite) => User.doesKeyBelongToInvite(key, invite)),
+        ),
+      );
   }
```

Comment:
Here is what I've understood. Let me know if I am wrong anywhere. <br>
Let's say there are two users `Bob` and `Pete`. `Bob` is already invited to an event which is happening between 4:30-5:30. Whereas `Pete` is not invited to any event.

Now I am trying to create an event between 4:00-5:00. and I want to check availability of `Bob` and `Pete`
So the keys array would be 
```
keys = [
  {
    user_id: 1,
    dateTimeRange: {
      start_at: '4:00',
      end_at: '5:00'
    },
    excludeEvents: []
  },
  {
    user_id: 2,
    dateTimeRange: {
      start_at: '4:00',
      end_at: '5:00'
    },
    excludeEvents: []
  }
]
```
`user_id = 1` for pete and 2 for bob

So after the data is fetched from the database, in line number 103, here is what invites array will be looking like
```
invites = [
  {
    user_id: 2,
    //other stuff here
    event: {
      // other stuff here
      start_at: '4:30',
      end_at: '5:30'
    }
  }
]
```

now we want to reorder the invites array and make a new array which will be returned such that `returned array.length === keys.length` and order of `returned array` is same as that of keys.

So first we will check if keys[0] (Pete) is being invited to any other event. If he is invited to any other event ===> he is not available. Hence the negation sign.
`returned array[0] = true` (he is not invited to any event)

For keys[1] (Bob), we have to check if there is any invite for `Bob` in `invites` array. If there is ===> he is unavailable.

But if we do deep equal here, we can see that `Bob's` key is something like 
```
  {
    user_id: 2,
    dateTimeRange: {
      start_at: '4:00',
      end_at: '5:00'
    },
    excludeEvents: []
  }

```

whereas his invite is something like this

```
{
    user_id: 2,
    //other stuff here
    event: {
      // other stuff here
      start_at: '4:30',
      end_at: '5:30'
    }
  }
```

So I don't think we can do deep equal because.
1. The structure of `key` and `invite` are different (but this can easily be modified)
2. `DateTimeRange` of the key is not equal to `event.start_at` and `event.end_at`. `Deep-Equal` will return false here. But it should be returning truee. Hence i made that `doesKeyBelongToInvite` function

Reply:
Can't you simply convert the resultant invite item to match the structure of the key and then equate.